internal/impl/mysql/input_mysql_stream.go

```diff
@@ -0,0 +1,637 @@
+			Table:     e.Table.Name,
+			Position:  i.currentLogPosition,
+		}
+	}
+
+	return nil
+}
+
+func (i *mysqlStreamInput) syncBinlogPosition(ctx context.Context) error {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if i.currentLogPosition == nil {
+		i.logger.Warn("No current bingLog position")
+		return errors.New("no current binlog position")
+	}
+
+	var (
+		positionInByte []byte
+		err            error
+	)
+	if positionInByte, err = json.Marshal(*i.currentLogPosition); err != nil {
+		i.logger.Errorf("Failed to marshal binlog position: %v", err)
+		return err
+	}
+
+	var cErr error
+	if err := i.res.AccessCache(ctx, i.binLogCache, func(c service.Cache) {
+		cErr = c.Set(ctx, binLogCacheKey, positionInByte, nil)
+		if cErr != nil {
+			i.logger.Errorf("Failed to store binlog position: %v", cErr)
+		}
+	}); err != nil {
+		i.logger.Errorf("Access cache error %v", err)
+		return err
+	}
+
+	return cErr
+}
+
+func (i *mysqlStreamInput) ReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
+	i.mutex.Lock()
+	msgChan := i.msgChan
+	i.mutex.Unlock()
+	if msgChan == nil {
+		return nil, nil, service.ErrNotConnected
+	}
+
+	select {
+	case m, open := <-msgChan:
+		if !open {
+			return nil, nil, service.ErrNotConnected
+		}
+		return m.msg, m.ackFn, nil
+	case <-i.shutSig.HasStoppedChan():
+		return nil, nil, service.ErrNotConnected
+	case <-ctx.Done():
+	}
+
+	return nil, nil, ctx.Err()
+}
+
+func (i *mysqlStreamInput) Close(ctx context.Context) error {
+	i.shutSig.TriggerSoftStop()
+	select {
+	case <-ctx.Done():
+	case <-time.After(shutdownTimeout):
+	case <-i.shutSig.HasStoppedChan():
+	}
+	i.shutSig.TriggerHardStop()
+	if i.canal != nil {
+		i.canal.SyncedPosition()
+		i.canal.Close()
+	}
+
+	return i.snapshot.releaseSnapshot(ctx)
+}
+
+// ---- Redpanda Connect specific methods end----
+
+// --- MySQL Canal handler methods ----
+
+func (i *mysqlStreamInput) OnRotate(eh *replication.EventHeader, re *replication.RotateEvent) error {
+	i.mutex.Lock()
+	flushedBatch, err := i.batchPolicy.Flush(i.streamCtx)
+	if err != nil {
+		i.logger.Debugf("Flush batch error: %w", err)
+		return err
+	}
+
+	if ok := i.flushBatch(i.streamCtx, i.cp, flushedBatch, i.currentLogPosition); !ok {
+		return errors.New("failed to flush batch")
+	}
+
+	i.currentLogPosition.Pos = uint32(re.Position)
+	i.currentLogPosition.Name = string(re.NextLogName)
+	i.mutex.Unlock()
+
+	return nil
+}
+
+func (i *mysqlStreamInput) OnRow(e *canal.RowsEvent) error {
+	if _, ok := i.tablesFilterMap[e.Table.Name]; !ok {
+		return nil
+	}
+
+	switch e.Action {
+	case canal.InsertAction:
+		return i.onMessage(e, 0, 1)
+	case canal.DeleteAction:
+		return i.onMessage(e, 0, 1)
+	case canal.UpdateAction:
+		return i.onMessage(e, 1, 2)
+	default:
+		return errors.New("invalid rows action")
+	}
+}
+
+func (i *mysqlStreamInput) OnPosSynced(eh *replication.EventHeader, pos mysqlReplications.Position, gtid mysqlReplications.GTIDSet, synced bool) error {
```

Comment:
Can you explain when this is called? Because right now we seem to blindly overwrite `currentLogPosition` in multiple places and I don't know if that's always OK to do.